src/transformers/models/smolvlm/modular_smolvlm.py

```diff
@@ -0,0 +1,333 @@
+            image_positions = (cur_ids == self.image_token_id).nonzero(as_tuple=True)[0]
+            num_image_tokens = len(image_positions)
+
+            # If no <image> => text-only
+            if num_image_tokens == 0:
+                # We do not consume any row from image_hidden_states;
+                # but we do a zero-length slice so the image encoder is in the graph.
+                empty_slice = image_hidden_states[0][:0, :]  # shape (0, D)
+                # Concatenate text plus that empty slice.
+                # NOTE: this is important for DeepSpeed.
+                merged_text_only = torch.cat([cur_embeds, empty_slice], dim=0)
+                merged_outputs.append(merged_text_only)
+                continue
+
+            # Otherwise, we have at least one <image> token.
+            # Typically, if each image is S embeddings, we expect the total # of <image> tokens
+            # in this sample to be multiple of S => each group of S tokens = 1 image
+            if num_image_tokens % S != 0:
+                raise ValueError(
+                    f"Sample {b_idx} has {num_image_tokens} <image> tokens, not a multiple of S={S}. "
+                    "Cannot map them to blocks of shape (S, D)."
+                )
+
+            # We'll chunk image_positions into groups of size S
+            positions_list = image_positions.tolist()
+            # Example: if num_image_tokens=162 and S=81 => we have 2 images => 2 chunks each of length 81
+            chunks = [positions_list[i : i + S] for i in range(0, num_image_tokens, S)]
+
+            # We'll build a list of segments: text, then image row(s), text, etc.
+            segments = []
+            text_start = 0
+
+            # For each chunk (each chunk => 1 image)
+            for chunk in chunks:
+                # image_hidden_states[image_offset] => shape (S, D)
+                cur_block = image_hidden_states[image_offset]
+                image_offset += 1
+
+                # We'll iterate over the S positions in ascending order
+                for i_s, pos in enumerate(chunk):
+                    # Add text from [text_start..pos)
+                    if pos > text_start:
+                        segments.append(cur_embeds[text_start:pos])
+                    # Then add one row from cur_block => shape (1, D)
+                    row_of_block = cur_block[i_s : i_s + 1, :]
+                    segments.append(row_of_block)
+                    # skip the <image> token
+                    text_start = pos + 1
+
+            # leftover text after the final <image> token
+            if text_start < T:
+                segments.append(cur_embeds[text_start:])
+
+            # cat them into a single (T_b, D) tensor
+            merged_sample = torch.cat(segments, dim=0)
+            merged_outputs.append(merged_sample)
+
+        merged_outputs = torch.stack(merged_outputs)
+        # assert (old_merger_outputs==merged_outputs).all()
+        return merged_outputs
```

Comment:
I'd be pro-removing this and replacing it with a vectorized merger, I think a prefix sum can keep track of where the images are - I drafted something, I'm not sure it matches all cases but let's see

```suggestion
    batch_size, seq_length, hidden_size_text = inputs_embeds.shape
    num_image_blocks, patch_size, hidden_size_img = image_hidden_states.shape

    image_mask = (input_ids == image_token_id)
    num_image_tokens = image_mask.sum(dim=1)

    blocks_per_sample = num_image_tokens // patch_size
    
    offsets = F.pad(blocks_per_sample.cumsum(dim=0), (1, 0), value=0)
    block_offset = offsets[:-1]
    # I think something like this should work and separate chunks cleanly
    # TODO verify but for sure there's a low-hanging vectoriztion fruit
    row_cum = image_mask.cumsum(dim=-1)
    chunk_idx = (row_cum - 1) // patch_size
    local_idx = (row_cum - 1) % patch_size
    block_idx = block_offset.unsqueeze(1) + chunk_idx

    merged_outputs = inputs_embeds.clone()
    merged_outputs[image_mask] = image_hidden_states[block_idx[image_mask], local_idx[image_mask], :]

```

Reply:
This is not deepspeed-compatible, which is why we altered the original inputs merger